File: 1677-matrix-diagonal-sum/matrix-diagonal-sum.java
class Solution {
    public static int diagonalSum(int[][] mat) {
        int totalSum = 0;

        for(int i = 0; i < mat.length; i++){
            totalSum += mat[i][i] + mat[i][mat.length - i - 1];
        }

        if(mat.length % 2 != 0){
            totalSum -= mat[mat.length / 2][mat.length / 2];
        }

        return totalSum;
    }
}

File: 1677-matrix-diagonal-sum/matrix-diagonal-sum.scala
object Solution {
    def diagonalSum(mat: Array[Array[Int]]): Int = {
        def iter(matrix: Array[Array[Int]], index: Int, sum: Int): Int =
            matrix.headOption match
                case Some(column) =>
                    val leftAndRightSum = column(index) + column(mat.head.length - index - 1)
                    iter(matrix.tail, index + 1, sum + leftAndRightSum)
                case None         =>
                    if(mat.length % 2 == 0) then sum
                    else sum - mat(mat.length / 2)(mat.head.length / 2)

        iter(mat, 0, 0)
    }
}